existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,292 @@
+Compiler is smart enough to choose appropriate types.
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
+Let's take a look how this is done in datatypes.
+You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.
+
+```haskell
+data Elem = forall a. Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+```
+
+This is a heterogeneous list that contains values of different types.
+You can't do that with an universally quantified list because every element in that list would need to have the same type.
+But, there are not many things you can do with elements of this list.
+Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.
+However, adding constraints to the `a` variable may improve the situation:
+
+```haskell
+-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
+data Elem = forall a. (Show a) => Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+
+-- We can use 'print' here because inner types of 'Elem' has 'Show' constraint
+printElem :: Elem -> IO ()
+printElem (Elem x) = print x
+
+main :: IO ()
+main = fmap printElem multitypedList
+
+λ> "a"
+λ> 1
+λ> Just 5
+```
+Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.
+
+```haskell
+showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
+showFunc (x:xs) => (show x) ++ (showFunc xs)
+```
+
+Other useful example of hiding type variables is `SomeException` wrapper
+
+```haskell
+data SomeException = forall a. Exception a => SomeException a
+```
+
+It allows you to catch exceptions of any type.
+As long, as you hadn't specified its type, (and it is not obvious to the compiler, from the handler for example), all exceptions will be caught, wrapped in `SomeException` datatype.
+
+However, again, there is a limitation.
+We know nothing about inner exception.
+Yes, we can for example print it, but sometimes we need more information.
+We can't use type inheritance to describe, what type we want to get
+
+```haskell
+-- Something like that won't work.
+-- we know to little about type inside,
+-- to be sure that it is 'MyException'
+myFromException :: SomeException -> MyException
+myFromException (SomeException (ex :: MyException)) = ex
+```
+
+Luckily for users, module `Control.Exception` provides special function `fromException`, the purpose of which is to extract inner variable and cast to specified type. How it is done on the inside, is a topic of a different article.
```

Comment:
>special function `fromException`

AFAIU there is nothing special about this `fromException`. 

>inner variable

Better say 'inner value', or 'wrapped value'.